### PR TITLE
Building customer class and tests.

### PR DESCRIPTION
**Pull request by pJanks**
### What's this PR do?
This adds catches to my fetch calls. It also builds out all or most of the tests I need for the Customer class and passes them. I am testing all default properties and methods. I am currently having trouble trying to figure out how to check the date property though.
### Where should the reviewer start?
I really want feedback on my Customer-tests with emphasis on if they are robust enough and if they thoroughly test method paths. I would also like feedback on things that may be refactored in the Customer class
### How should this be manually tested?
Review the files changed and pull down the branch to make sure tests pass and there aren't any bugs.
### Potential changes?
My next step is to build out a child class of Customer for Manager and write the tests to get Manger's exclusive tests and properties tested.

@khalidwilliams
@hannahhch

### CODE REVIEW

File: test/Customer-test.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  beforeEach(() => {
+    date = new Date();
+    currentDate = date.getFullYear() + "/" + (date.getMonth() + 1) + "/" + date.getDate();
+    finalDate = new Date(currentDate)
+    customer = new Customer({
+      id: 1,
+      name: "Leatha Ullrich",
+      currentDate: finalDate,
+    })
+    rooms = [{
```
Comment (KevDev90) [inline test/Customer-test.js]: Something you could possibly consider.. instead of doing this sample data here before every test, you can make a separate directory called data and create a user-data, manager-data file etc and import them in here to make it look a little cleaner.

File: test/Customer-test.js [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        }
+      ])
+    })
+
+    it('should respond with a message if there are no rooms of that type', () => {
+      expect(customer.sortRoomByType('foo', rooms)).to.equal('We do not have that room type. I fiercely apologize!')
+    })
+  })
+
+  describe('bookRoom', () => {
```
Comment (KevDev90) [inline test/Customer-test.js]: Nice job busting this out into its own describe block and testing the different possible outcomes of the method.

File: test/Customer-test.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+      {
+        id: "5fwrgu4i7k55hl6t8",
+        userID: 14,
+        date: "2020/02/05",
+        roomNumber: 5,
+        roomServiceCharges: []
+      }
+    ]
+  })
+  describe('default propertiex', () => {
```
Comment (KevDev90) [inline test/Customer-test.js]: Awesome job overall man! Nice job testing your happy and sad paths!

File: test/Customer-test.js (hunk 1 above)
Comment (pJanks) [reply test/Customer-test.js]: That is actually a great idea. I have been feeling like hard coding it in probably isn't the best way.

File: test/Customer-test.js (hunk 3 above)
Comment (pJanks) [reply test/Customer-test.js]: Thank you. I hope I covered my bases as well as I think I did. We will see once I start working with the whole dataset instead of the hardcoded one.

File: test/Customer-test.js (hunk 2 above)
Comment (pJanks) [reply test/Customer-test.js]: Thank you

File: src/Customer.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+class Customer {
+  constructor(user) {
+    this.id = user.id;
+    this.name = user.name;
+    this.customerBookings = [];
+    this.date = user.currentDate;
+    this.unavailableRoomsByDate = []
+  }
+
+  getCurrentDate() {
```
Comment (khalidwilliams) [inline src/Customer.js]: This method is cool! It makes sense to use in the browser, but for tests you shouldn't rely on it (because its value will change every day).

Make sure you unit test it somewhere in `Customer-test`, then just pass in a hard coded date when testing methods that takes its output as an argument.

File: src/Customer.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        })
+        return acc;
+      }, []);
+      this.availableRoomsByDate = this.availableRoomsByDate.reduce((acc, availableRoom) => {
+        bookings.forEach(booking => {
+          if (booking.roomNumber === availableRoom.number && ((Math.abs(today) - Math.abs(new Date(booking.date))) < 1)) {
+            acc.push(availableRoom)
+          } else if (booking.roomNumber === availableRoom.number && ((Math.abs(today) - Math.abs(new Date(booking.date))) > 1)) {
+            this.unavailableRoomsByDate.push(availableRoom)
+          }
```
Comment (khalidwilliams) [inline src/Customer.js]: I'm a little confused by the second reduce here, the logic in this method is hard to follow. Are you cross referencing bookings to prevent duplication? Or to avoid skipping rooms?

File: test/Customer-test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: Thank you I have fixed that.

File: test/Customer-test.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+    // it('should know the current date', () => {
+    //   expect(customer.date).to.equal(Date.now());
```
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: I believe I was able to solve this problem in the exact way your are describing.

File: test/Customer-test.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+      newBookings = [{
+        number: 3,
+        roomType: "single room",
+        bidet: false,
+        bedSize: "king",
+        numBeds: 1,
+        costPerNight: 491.14
+      }]
+      expect(customer.checkAvailableDates(newBookings, newRooms, '2020/01/12',
+        customer.getCurrentDate())).to.equal('There are no rooms available for that day! I fierecly apologize!')
```
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: Yes, I should have thought about this in the first place. I had to go back and hardcode a date because a lot of tests started failing when the date changed and I was using the live date.

File: test/Customer-test.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    it('should allow a user to select an available room', () => {
+      customer.checkAvailableDates(bookings, rooms, '2020/01/16',
+        customer.getCurrentDate())
+      customer.bookRoom(3, "2020/01/16")
+      expect(customer.customerBookings.length).to.equal(1)
+    })
+
+    it('should not allow a user to select an unavailable room', () => {
+      customer.checkAvailableDates(bookings, rooms, '2020/01/12',
+        customer.getCurrentDate())
```
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: Thank you!

File: test/Customer-test.js [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+      expect(customer.unavailableRoomsByDate.length).to.equal(2)
+    })
+
+    it('should keep track of rooms that are available on a date', () => {
+      customer.checkAvailableDates(bookings, rooms, '2020/01/12',
+        customer.getCurrentDate())
+      expect(customer.availableRoomsByDate.length).to.equal(3)
+    })
```
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: Yes. As I said earlier I have fixed it now. Thank you

File: test/Customer-test.js [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    it('sould be able to see all bookings made', () => {
+      customer.viewBookings(bookings);
+      expect(customer.customerBookings.length).to.equal(3);
+    })
+  });
+
+  describe('calculateMoneySpent', () => {
+    it('should be able to see total cost spent on rooms', () => {
+      customer.calculateMoneySpent(bookings, rooms);
+      expect(customer.totalMoneySpent).to.equal(358.4)
```
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: I believe I have since fixed this thank you.

File: test/Customer-test.js [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+          bidet: true,
+          bedSize: "queen",
+          numBeds: 2,
+          costPerNight: 340.17
+        }
+      ])
+    })
+
+    it('should respond with a message if there are no rooms of that type', () => {
+      expect(customer.sortRoomByType('foo', rooms)).to.equal('We do not have that room type. I fiercely apologize!')
```
[the remark this answers is not in the document]
Comment (pJanks) [reply test/Customer-test.js]: Thank you!

File: src/Customer.js (hunk 4 above)
Comment (pJanks) [reply src/Customer.js]: Fixed! Thank you!

File: src/Customer.js (hunk 5 above)
Comment (pJanks) [reply src/Customer.js]: I still haven't figured out what I am doing. This function works but If I take out one of the reduces it breaks everything. I did think about it before that it seems redundant, but if you have any tips on how to refactor it to work better they would be appreciated.

**Comment by pJanks** [comment]
@khalidwilliams There was only one thing that I haven't been able to fix so far. If you could review the comment I left on the double reduce function and maybe give a little feedback. I know it's hard to follow, that's one of the reasons it's hard for me to refactor it. I know it works, but I'm not 100% sure why it works.

File: src/Customer.js (hunk 5 above)
Comment (khalidwilliams) [reply src/Customer.js]: I'm not sure, it seems like you're cross referencing the bookings. I don't have time to dive into how you're passing data into here. You could maybe split the reduces into their own helper functions, but it doesn't need to be a priority if you're not done with the other user stories.

Comment (pJanks) [reply src/Customer.js]: Thanks. I was at least able to chain the reduce and remove the redundancy of reassigning the class property.
